relayer/src/kill_switch.rs

```diff
@@ -0,0 +1,301 @@
+use std::{
+    path::Path,
+    process,
+    time::{Duration, Instant},
+};
+
+use block_finality_archiver::BlockFinalityProofWithHash;
+use ethereum_client::{EthApi, MerkleRootEntry, TxHash, TxStatus};
+use gear_rpc_client::GearApi;
+use parity_scale_codec::Decode;
+use prover::proving::GenesisConfig;
+
+use crate::{
+    common::{submit_proof_to_ethereum, sync_authority_set_id},
+    proof_storage::ProofStorage,
+    prover_interface::{self, FinalProof},
+};
+
+mod block_finality_archiver;
+
+const MIN_MAIN_LOOP_DURATION: Duration = Duration::from_secs(12);
+
+enum State {
+    Normal,
+    // Before exit we need to wait for the kill switch transaction to be finalized.
+    WaitingForKillSwitchTxFin { tx_hash: TxHash, proof: FinalProof },
+}
+
+pub struct KillSwitchRelayer {
+    gear_api: GearApi,
+    eth_api: EthApi,
+    genesis_config: GenesisConfig,
+    proof_storage: Box<dyn ProofStorage>,
+
+    // Next eth block to process.
+    from_eth_block: Option<u64>,
```

Comment:
In fact its not a `latest_eth_block` because it used as start of block range when fetching merkle root events.

```rust
        let events = self
            .eth_api
            .fetch_merkle_roots_in_range(from_eth_block, last_finalized_block)
            .await?;
```

Renamed `from_eth_block` -> `start_from_eth_block` for better clarity.